WIP : Refactor for AMDFlang

PaulMullowney opened the pull request:
This is a sequence of commits to get things working well with AMD Flang compiler (AFAR drop 6.2) and OMP Target support. I have tested single device on MI250 and MI300X and get good results out to Truncation 639 (single and double precision).

Ultimately, I wouldn't worry about the performance reported below as most of the time is spent in the D2H H2D copies in TRGTOL and TRLTOG for MI300X. I am guessing the same is true for H100. I think setups with fast CPU-GPU transfer such as an MI300A and/or a GH200 will show much better speeds. This should be verified.

The runs below show that correctness is preserved:
ectrans-dev/build/bin/ectrans-benchmark-gpu-sp (or dp) --niter 20 --truncation 639 --nlev 137 --nfld 1 --vordiv --scders --uvders --norms

**GRAPHS (FFT, BLAS) ON:
MI300X 639:**

Single:
time step      1 took 33.3881 | zspvor max err= 0.134E-06 | zspdiv max err= 0.134E-06 | zspsc3a max err= 0.219E-06 | zspsc2 max err= 0.119E-06
time step      2 took  0.4045 | zspvor max err= 0.358E-06 | zspdiv max err= 0.303E-06 | zspsc3a max err= 0.472E-06 | zspsc2 max err= 0.238E-06
time step      3 took  0.3982 | zspvor max err= 0.477E-06 | zspdiv max err= 0.477E-06 | zspsc3a max err= 0.809E-06 | zspsc2 max err= 0.477E-06

Double:
time step      1 took 42.8045 | zspvor max err= 0.833E-14 | zspdiv max err= 0.833E-14 | zspsc3a max err= 0.544E-14 | zspsc2 max err= 0.544E-14
time step      2 took  0.6729 | zspvor max err= 0.170E-13 | zspdiv max err= 0.170E-13 | zspsc3a max err= 0.103E-13 | zspsc2 max err= 0.103E-13
time step      3 took  0.8979 | zspvor max err= 0.246E-13 | zspdiv max err= 0.246E-13 | zspsc3a max err= 0.159E-13 | zspsc2 max err= 0.159E-13

**H100 639:** 

Single:
time step      1 took 19.1471 | zspvor max err= 0.119E-06 | zspdiv max err= 0.119E-06 | zspsc3a max err= 0.119E-06 | zspsc2 max err= 0.119E-06
time step      2 took  1.2367 | zspvor max err= 0.238E-06 | zspdiv max err= 0.119E-06 | zspsc3a max err= 0.119E-06 | zspsc2 max err= 0.119E-06
time step      3 took  1.2350 | zspvor max err= 0.358E-06 | zspdiv max err= 0.119E-06 | zspsc3a max err= 0.238E-06 | zspsc2 max err= 0.238E-06

Double:
GPU runtime error 5: CUFFT_INTERNAL_ERROR
Out of Memory??
Runs fine for truncation 319

**GRAPHS (FFT, BLAS) OFF:
MI300X 639:**

Single:
time step      1 took 35.4651 | zspvor max err= 0.134E-06 | zspdiv max err= 0.134E-06 | zspsc3a max err= 0.219E-06 | zspsc2 max err= 0.119E-06
time step      2 took  0.3948 | zspvor max err= 0.358E-06 | zspdiv max err= 0.303E-06 | zspsc3a max err= 0.472E-06 | zspsc2 max err= 0.238E-06
time step      3 took  0.4010 | zspvor max err= 0.477E-06 | zspdiv max err= 0.477E-06 | zspsc3a max err= 0.809E-06 | zspsc2 max err= 0.477E-06

Double:
time step      1 took 45.6732 | zspvor max err= 0.833E-14 | zspdiv max err= 0.833E-14 | zspsc3a max err= 0.544E-14 | zspsc2 max err= 0.544E-14
time step      2 took  0.6735 | zspvor max err= 0.170E-13 | zspdiv max err= 0.170E-13 | zspsc3a max err= 0.103E-13 | zspsc2 max err= 0.103E-13
time step      3 took  0.8981 | zspvor max err= 0.246E-13 | zspdiv max err= 0.246E-13 | zspsc3a max err= 0.159E-13 | zspsc2 max err= 0.159E-13

**H100 639:** 

Single:
time step      1 took 18.8916 | zspvor max err= 0.119E-06 | zspdiv max err= 0.119E-06 | zspsc3a max err= 0.119E-06 | zspsc2 max err= 0.119E-06
time step      2 took  1.2467 | zspvor max err= 0.238E-06 | zspdiv max err= 0.119E-06 | zspsc3a max err= 0.119E-06 | zspsc2 max err= 0.119E-06
time step      3 took  1.2443 | zspvor max err= 0.358E-06 | zspdiv max err= 0.119E-06 | zspsc3a max err= 0.238E-06 | zspsc2 max err= 0.238E-06

Double:
 CUFFT_INTERNAL_ERROR
 Out of Memory??
 Runs fine for truncation 319!

samhatfield commented:
Thanks for this @PaulMullowney. Paul and I will discuss offline about testing this branch on an internal AMD server.

samhatfield commented on src/trans/gpu/internal/ltdir_mod.F90:
```diff
@@ -78,7 +78,7 @@ SUBROUTINE LTDIR(ALLOCATOR,HLTDIR,ZINPS,ZINPA,ZINPS0,ZINPA0,KF_FS,KF_UV,KF_SCALA
     USE TPM_STATS,              ONLY: GSTATS => GSTATS_NVTX
     USE BUFFERED_ALLOCATOR_MOD, ONLY: BUFFERED_ALLOCATOR, ASSIGN_PTR, GET_ALLOCATION
-    USE ISO_C_BINDING,          ONLY: C_SIZE_T, C_F_POINTER, C_LOC, C_SIZEOF
+    USE ISO_C_BINDING,          ONLY: C_SIZE_T, C_F_POINTER, C_LOC, C_SIZEOF, C_INT8_T
```
```suggestion
    USE ISO_C_BINDING,          ONLY: C_SIZE_T, C_F_POINTER, C_LOC, C_SIZEOF
```
Not used here I think?

samhatfield commented on src/trans/gpu/algor/buffered_allocator_mod.F90:
```diff
@@ -121,18 +121,20 @@ FUNCTION GET_ALLOCATION(ALLOCATOR, RESERVATION)
+    IF (END_IN_BYTES > SIZE(SRC, KIND=C_SIZE_T)) THEN
       CALL ABORT_TRANS("Logical Error in ASSIGN_PTR - OOB assignment")
-    ENDIF
+   ENDIF
```
```suggestion
    ENDIF
```
Indentation typo.

samhatfield replied:
Still there. BTW Github's "add suggestion" buttons there make it super easy to fix these annoying tiny changes.

samhatfield commented on src/trans/gpu/algor/hicblas_gemm.hip.cpp:
```diff
@@ -124,28 +124,21 @@ void run_group_graph(Gemm &&gemm, int resol_id, int m, const int *n,
-      gemm(stream, m, n[i], k[i], alpha, A + offsetsA[i], lda, B + offsetsB[i],
+      gemm(captureStream, m, n[i], k[i], alpha, A + offsetsA[i], lda, B + offsetsB[i],
            ldb[i], beta, C + offsetsC[i], ldc);
-      hipGraph_t my_graph;
-      HIC_CHECK(hipStreamEndCapture(stream, &my_graph));
-      hipGraphNode_t my_node;
-      HIC_CHECK(
-          hipGraphAddChildGraphNode(&my_node, new_graph, nullptr, 0, my_graph));
```
Again it looks like these changes the graph capturing behaviour quite a lot. Is this strictly necessary for amdflang?

samhatfield commented on src/trans/gpu/internal/trltom_mod.F90:
```diff
@@ -189,7 +189,7 @@ SUBROUTINE TRLTOM(ALLOCATOR,HTRLTOM,PFBUF_IN,PFBUF,KF_FS)
 #ifdef USE_GPU_AWARE_MPI
 #ifdef OMPGPU
-      !$OMP TARGET DATA USE_DEVICE_PTR(PFBUF_IN,PFBUF)
+      !$OMP TARGET DATA USE_DEVICE_ADDR(PFBUF_IN,PFBUF)
```
What's the need for this change?

samhatfield commented:
I had a chat with Paul and we agreed that https://github.com/ecmwf-ifs/ectrans/pull/291/commits/0ff6e306911cafb1f505b65dcb1e78a98b9dc470 should be considered in a separate PR, as it's not about "getting it working on MI300" but rather about optimisation. Is that right @PaulMullowney?

samhatfield commented:
Benchmark of this branch on MI300A compared with other platforms:

<img width="1555" height="738" alt="ectrans-single-node-comparison-25-07-2025" src="https://github.com/user-attachments/assets/e370241b-4e1f-4ecb-a3af-4e1a49273237" />

1 MI300A matches 4 MI250Xs.

samhatfield commented on src/trans/gpu/algor/hicblas_gemm.hip.cpp:
```diff
@@ -171,6 +170,12 @@ void run_group(Gemm &&gemm, int resol_id, int m, const int *n, const int *k,
     gemm(stream, m, n[i], k[i], alpha, A + offsetsA[i], lda, B + offsetsB[i],
          ldb[i], beta, C + offsetsC[i], ldc);
   }
+#ifdef CUDAGPU
+  HIC_CHECK(cudaStreamSynchronize(stream));
+#endif
+#ifdef HIPGPU
+  HIC_CHECK(hipStreamSynchronize(stream));
+#endif
```
This `#ifdef` shouldn't be necessary if you add
```c
#define hipStreamSynchronize cudaStreamSynchronize
```
to [this file](https://github.com/PaulMullowney/ectrans/blob/amdflang-rework/src/trans/gpu/algor/hicblas_cuda.h). Same for the code above and in hicfft.hip.cpp.

samhatfield commented:
> Sure. Anything you can do to move this along is much appreciated. I am on PTO at the moment, but I will respond with some delays.

Ah no problem. I prepared the branch, then realised it was kinda pointless as there are only a couple of changes I propose above. When you're back, I suggest you:
1) split out commits related to graph optimisation into a separate PR
2) apply the suggested changes above to buffered_allocator_mod.F90 and ltdir_mod.F90
3) apply the suggested changes above to hicblas_gemm.hip.cpp in the new PR

Then we will get the two over the line separately. No rush of course.

PaulMullowney commented:
> > Having rebased against master, performance has dropped by about 10%. I don't know if this is expected or not.
> 
> You mean develop? I think your local checkout of develop needs to be updated because [amdflang-rework's history](https://github.com/PaulMullowney/ectrans/commits/amdflang-rework/) diverges from [develop's history](https://github.com/ecmwf-ifs/ectrans/commits/develop) after around July 7.

Yes. I meant develop. I did update my local changes to develop. I am gonna try CCE for comparison.

samhatfield commented:
> Yes. I meant develop. I did update my local changes to develop. I am gonna try CCE for comparison.

But the histories of your branch and develop diverge, so your local checkout must be out of date. Your branch is missing many commits from develop, including [this one](https://github.com/ecmwf-ifs/ectrans/commit/7c90a65879aceaf1ad09c93bf812c1350a2ad3c0). Can you check again?

PaulMullowney commented:
> > Yes. I meant develop. I did update my local changes to develop. I am gonna try CCE for comparison.
> 
> But the histories of your branch and develop diverge, so your local checkout must be out of date. Your branch is missing many commits from develop, including [this one](https://github.com/ecmwf-ifs/ectrans/commit/7c90a65879aceaf1ad09c93bf812c1350a2ad3c0). Can you check again?

Locally, I see:

commit cf68a2cb0d9137210927581b0f965f9a6730562f
Author: Paul Mullowney <Paul.Mullowney@amd.com>
Date:   Wed Jul 9 15:24:34 2025 -0500

    AMDFlang does not well support array assignment. Ultimately this will be handled with a work distribute clause but for now, we should do loops.

commit 7c90a65879aceaf1ad09c93bf812c1350a2ad3c0 (origin/develop, develop)
Author: PaulMullowney <60452402+PaulMullowney@users.noreply.github.com>
Date:   Mon Aug 11 10:46:51 2025 -0600

    Simplification of the graph implementation for FFT and BLAS. (#300)

samhatfield commented:
The SHA of 'AMDFlang does not well support..." is cf68a2c which doesn't match the one [online](https://github.com/PaulMullowney/ectrans/commit/16105ed3e2edc6ce5f0a971c5fbebc28fd33b9ea) (16105ed) so I'm not sure what's going on. Perhaps you need to force push amdflang-new to your Github remote.

samhatfield commented:
Also can you force push your amdflang-rework here? Then I can give it a try as well.

PaulMullowney commented:
> > I just rebased locally again. I am now seeing .3 seconds per time step for 639 where as I used to see .22. Everything is consistently worse across the board.
> > 102 LTINV_CTL - Inv. Legendre transform 20 23.641 23.641 0.30 0.00 107 FTINV_CTL - Inv. Fourier transform 20 71.438 71.438 0.92 0.00 140 SULEG - Comp. of Leg. poly. 1 481.353 481.353 0.31 0.00 152 LTINV_CTL - M to L transposition 20 27.768 27.768 0.36 0.00 153 LTDIR_CTL - L to M transposition 20 14.220 14.220 0.18 0.00 157 FTINV_CTL - L to G transposition 20 79.967 79.967 1.03 0.00 158 FTDIR_CTL - G to L transposition 20 32.793 32.793 0.42 0.00
> > I have another branch (amdflang-rework-old) where I can recover the performance. This does not have the rebase.
> 
> Good idea to keep that branch yeah, just in case. I think this performance drop can be explained: PR #122 changed the behaviour of the benchmark program. Specifically, `ldvorgp` and `lddivgp` are now `.true.` when `--vordiv` is passed. They used to be `.false.` permanently. What performance do you get if you hard-code `ldvorgp` and `lddivgp` to `.false.` in the benchmark program? Your numerics may be broken but you should recover the same performance.

Locally, I modified:
diff --git a/src/programs/ectrans-benchmark.F90 b/src/programs/ectrans-benchmark.F90
index cb48b57..daff3e9 100644
--- a/src/programs/ectrans-benchmark.F90
+++ b/src/programs/ectrans-benchmark.F90
@@ -629,13 +629,13 @@ do jstep = 1, iters+iters_warmup
   if (icall_mode == 1) then
     call inv_trans(pspvor=zspvor, pspdiv=zspdiv, pspscalar=zspscalar, pgp=zgp, &
       &            kvsetuv=ivset, kvsetsc=ivsetsc, &
-      &            ldscders=lscders, ldvorgp=lvordiv, lddivgp=lvordiv, lduvder=luvder, &
+      &            ldscders=lscders, ldvorgp=.false., lddivgp=.false., lduvder=luvder, &
       &            kproma=nproma)
   else
     call inv_trans(pspvor=zspvor, pspdiv=zspdiv, pspsc3a=zspsc3a, pspsc2=zspsc2, pgpuv=zgpuv, &
       &            pgp3a=zgp3a, pgp2=zgp2, &
       &            kvsetuv=ivset, kvsetsc2=ivsetsc2, kvsetsc3a=ivset, &
-      &            ldscders=lscders, ldvorgp=lvordiv, lddivgp=lvordiv, lduvder=luvder, kproma=nproma)
+      &            ldscders=lscders, ldvorgp=.false., lddivgp=.false., lduvder=luvder, kproma=nproma)


I see:

time step      2 took  0.2555 | zspvor max err= 0.213E+12 | zspdiv max err= 0.213E+12 | zspsc3a max err= 0.477E-06 | zspsc2 max err= 0.238E-06
time step      3 took  0.2365 | zspvor max err= 0.223E+17 | zspdiv max err= 0.223E+17 | zspsc3a max err= 0.715E-06 | zspsc2 max err= 0.358E-06
time step      4 took  0.2357 | zspvor max err= 0.495E+21 | zspdiv max err= 0.495E+21 | zspsc3a max err= 0.954E-06 | zspsc2 max err= 0.477E-06
time step      5 took  0.2515 | zspvor max err=       Inf | zspdiv max err=       Inf | zspsc3a max err= 0.119E-05 | zspsc2 max err= 0.596E-06
time step      6 took  0.2366 | zspvor max err=       Inf | zspdiv max err=       Inf | zspsc3a max err= 0.131E-05 | zspsc2 max err= 0.715E-06
time step      7 took  0.2350 | zspvor max err=       Inf | zspdiv max err=       Inf | zspsc3a max err= 0.155E-05 | zspsc2 max err= 0.715E-06

The L to G and G to L are slower but things are closer.

 102 LTINV_CTL      - Inv. Legendre transform       20        19.359        19.359      0.27      0.00
 107 FTINV_CTL      - Inv. Fourier transform        20        65.244        65.244      0.89      0.00
 140 SULEG          - Comp. of Leg. poly.            1       719.339       719.339      0.49      0.00
 152 LTINV_CTL      - M to L transposition          20        17.701        17.701      0.24      0.00
 153 LTDIR_CTL      - L to M transposition          20        13.818        13.818      0.19      0.00
 157 FTINV_CTL      - L to G transposition          20        30.572        30.572      0.42      0.00
 158 FTDIR_CTL      - G to L transposition          20        26.144        26.144      0.36      0.00

samhatfield commented:
PR is clean and is now tested properly on MI300.

<img width="1555" height="738" alt="ectrans-single-node-comparison-14-08-2025" src="https://github.com/user-attachments/assets/2e5baf69-15f9-47ef-9b01-167307a7681f" />

Good to merge.

samhatfield commented:
Merged.
